### PR TITLE
Backend: Add get orders (by user) endpoint

### LINKED ISSUE #99: Endpoint to get all orders for specific user
## User story
**Who:** **As a** developer
**What:** **I want to** get all orders made by specific user
**Why:** **so that I can** see it on Order summary page
## Acceptance criteria

The endpoint should show the order number, number of items, date of the order and the amount as shown below.

order-summary

## Implementation Details

Create the GET endpoint for 'api/orders?user={user_id}'

Read about Swagger here https://github.com/HackYourFuture-CPH/fp-class18/#-working-with-api-documentation---swagger and this article should also help you with that https://blog.logrocket.com/documenting-your-express-api-with-swagger/

### PR DESCRIPTION
**Pull request by JiuTak**
# Description

Please provide a short summary explaining what this PR is about.

Fixes #99 

# How to test?
localhost:3000/api/documentation

Check ORDERS 

# Checklist

- [x] I have performed a self-review of my own code
- [x] I have followed the name conventions for CSS Classnames and filenames, Components names and filenames, Style filenames, if you are in doubt check the the project README.MD and here https://github.com/HackYourFuture-CPH/curriculum/blob/master/review/review-checklist.md
- [x] I have commented my code, particularly in hard-to-understand areas, if you code was simple enough mark the box anyway
- [x] I have made corresponding changes to the documentation, if you code was simple enough mark the box anyway
- [x] This PR is ready to be merged and not breaking any other functionality

### CODE REVIEW

File: src/server/api/routes/orders.router.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
+/* TODO: This is just an example file to illustrate API routing and
+documentation. Can be deleted when the first real route is added. */
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js]: Nitpick: Feels like this comment doesn't apply anymore, can we remove it?

File: src/server/api/routes/orders.router.js [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+const express = require('express');
+
+const router = express.Router({ mergeParams: true });
+
+// controllers
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js]: Nitpick: 
The variable is called controller, and the file it imports from is called controller. 
How much more context is this comment adding? :)

File: src/server/api/routes/orders.router.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/* TODO: This is just an example file to illustrate API routing and
+documentation. Can be deleted when the first real route is added. */
+
+const express = require('express');
+
+const router = express.Router({ mergeParams: true });
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js:4]: Nitpick: I think it would be better if you grouped the imports (require) at the top, and put this line after the two imports!

**Review by nataliasowa — APPROVED** [review]
Tested and it works well in swagger. Since we both work on the different endpoints for orders we will probably need to push another PR for the review and align the select section as mine is a bit different than yours :)

File: src/server/api/routes/orders.router.js [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+ *      200:
+ *        description: Successful request
+ *      404:
+ *        description: incorrect entry with the provided id
+ *      5XX:
+ *        description: Unexpected error.
+ */
+router.get('/', (req, res, next) => {
+  ordersController
+    .getOrderByUserId(req.query.user)
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js]: Would it be more clear what data the endpoint expects if we renamed this query parameter to `user_id` instead?

File: src/server/api/routes/orders.router.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    .then((result) => {
+      if (result.length === 0) {
+        res.status(404).send('The user ID you provided does not exist');
+      } else {
+        return res.json(result);
+      }
+    })
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js]: I would move this logic into the controller, leaving the router to only focus on unpacking the HTTP request, calling the controller and then returning the result/error.

```
.then((result) => res.json(result))
```

File: src/server/api/routes/orders.router.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+ *        description: incorrect entry with the provided id
+ *      5XX:
+ *        description: Unexpected error.
+ */
+router.get('/', (req, res, next) => {
+  ordersController
+    .getOrderByUserId(req.query.user)
+    .then((result) => {
+      if (result.length === 0) {
+        res.status(404).send('The user ID you provided does not exist');
```
Comment (miltonhultgren) [inline src/server/api/routes/orders.router.js]: Could this also happen if the user does not have any orders?
How would we tell the difference?

File: src/server/api/controllers/orders.controller.js [hunk 7]
```diff
@@ -0,0 +1,33 @@
+const knex = require('../../config/db');
+const HttpError = require('../lib/utils/http-error');
+// const moment = require('moment-timezone');
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Nitpick: Remove commented out code!

File: src/server/api/controllers/orders.controller.js [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+  }
+  try {
+    const orders = await knex('orders')
+      .select(
+        'orders.user_id',
+        'orders.id as order_number',
+        'orders.updated_at',
+        knex.raw('sum( ?? * ?? ) as ??', ['quantity', 'price', 'total_amount']),
+      )
+
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Nitpick: This empty line confused me :D

File: src/server/api/controllers/orders.controller.js [hunk 9]
```diff
@@ -0,0 +1,33 @@
+const knex = require('../../config/db');
+const HttpError = require('../lib/utils/http-error');
+// const moment = require('moment-timezone');
+
+const getOrderByUserId = async (user_id) => {
+  if (isNaN(user_id)) {
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Be careful with `isNaN`, it only checks if `user_id` is equal to the global constant `NaN`. It does not check that `user_id` is null, undefined, or a string that is not a number. 
You'll need to extend this condition to catch those cases.

I'd suggest combining isNaN with parseInt (in HTTP all values are strings to being with).
```
const user_id = parseInt(query_user_id);
if (isNaN(userId)) {
```

I also see in this PR https://github.com/HackYourFuture-CPH/fp-class19/pull/174/files the author is checking that ID is higher than 0, did you agree on the business logic here?

File: src/server/api/controllers/orders.controller.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+      )
+
+      .leftJoin('order_items', 'order_items.order_id', '=', 'orders.id')
+      .leftJoin('users', 'users.id', '=', 'orders.user_id')
+      .leftJoin('products', 'products.id', '=', 'order_items.product_id')
+      .where({ user_id })
+      .groupBy('orders.id')
+      .count('orders.id as nr_of_items');
+
+    return orders;
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js:79]: How can we tell in this case if the orders list is empty because there is no user or the user has no orders? Does that matter for the end user of this API? 

In either case, perhaps it is most correct to throw an error in that case stating that no orders were found (since that we for sure know)?

File: src/server/api/controllers/orders.controller.js [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+  } catch (error) {
+    return error.message;
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Unpacking only the message here doesn't feel like it adds much value. Perhaps it would be better to throw a `HttpError` instead based on the error the database threw?
Something like:
```
} catch (error) {
  if (error instanceof HttpError) {
    throw error;
  }
  
  throw new HttpError('Something went wrong', 500);
}
```
This way, we don't leak information about what went wrong or what kind of database we are using.

**Review by miltonhultgren — CHANGES_REQUESTED** [review]
This PR needs a little bit more work. I'd look at the usage of `isNaN` and improving the error handling, specially in the case where no orders are found.

The comments marked with `nitpick` are fine to ignore if you want!

**Comment by miltonhultgren** [comment]
@NataliaS1994 @JiuTak How can you share the code for the selection part in a good way, think about what kind of function you could extract to import into both controllers for sharing that logic :)

**Comment by miltonhultgren** [comment]
I'd update the PR title and merge commit message to:
Add get orders (by user) endpoint

File: src/server/api/controllers/orders.controller.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+const knex = require('../../config/db');
+const HttpError = require('../lib/utils/http-error');
+// const moment = require('moment-timezone');
+
+const getOrderByUserId = async (user_id) => {
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Since this function returns more than one order if there are, maybe we should rename it to getOrdersByUserId?

File: src/server/api/controllers/orders.controller.js [hunk 13]
```diff
@@ -0,0 +1,37 @@
+const knex = require('../../config/db');
+const HttpError = require('../lib/utils/http-error');
+
+const getOrdersByUserId = async (raw_user_id) => {
+  const user_id = parseInt(raw_user_id, 10);
+
+  if (isNaN(user_id) || user_id < 1) {
+    throw new HttpError('The ID should be a number', 400);
```
Comment (Divyajg) [inline src/server/api/controllers/orders.controller.js:52]: response code 400 is not documented in the swagger responses.

**Review by Divyajg — COMMENTED** [review]
Looks good. response state 200 not changing for couple of wrong query inputs and total price returning null for orders having more than 1 product.

File: src/server/api/controllers/orders.controller.js (hunk 13 above)
Comment (JiuTak) [reply src/server/api/controllers/orders.controller.js:52]: image
👍 check please

and the I get all the orders and cancelled order showed null price, so I put order status so that we know why the price is null. 

Is that make sense? If I can improve in other ways, please let me know so that I can implement better code ☺️

File: src/server/api/controllers/orders.controller.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+      .leftJoin('users', 'users.id', '=', 'orders.user_id')
+      .leftJoin('products', 'products.id', '=', 'order_items.product_id')
+      .where({ user_id })
+      .groupBy('orders.id')
+      .count('orders.id as nr_of_items');
+    if (orders.length === 0) {
+      throw new HttpError('The user ID you provided does not exist', 404);
+    }
+
+    return new HttpError('The user have no orders yet', 404);
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]:
```suggestion
    if (orders.length === 0) {
      throw new HttpError('The user have no orders yet', 404);
    }

    return orders;
```

This seems like a little edit mistake, let's fix it and merge :)

**Review by miltonhultgren — APPROVED** [review]
Looks good to me 👍🏼 

Just one little mistake to fix, push a quick fix then you can merge :)

File: src/server/api/controllers/orders.controller.js (hunk 14 above)
Comment (JiuTak) [reply src/server/api/controllers/orders.controller.js]: Check updated one please😌

File: src/server/api/controllers/orders.controller.js [hunk 15]
```diff
@@ -0,0 +1,46 @@
+const knex = require('../../config/db');
+const HttpError = require('../lib/utils/http-error');
+
+const getOrdersByUserId = async (raw_user_id) => {
+  const user_id = +raw_user_id;
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js:49]: Sorry, but why was this changed again? 
Using the `+` operator to convert between data types is not a good practice.

File: src/server/api/controllers/orders.controller.js [hunk 16]
```diff
@@ -0,0 +1,46 @@
...
+
+  if (isNaN(user_id) || user_id < 1) {
+    throw new HttpError('The ID should be a number', 400);
+  }
+  try {
+    const user = await knex('users').select('id').where('id', '=', user_id);
+
+    if (user.length === 0) {
+      throw new HttpError('The user ID you provided does not exist', 404);
+    }
```
Comment (miltonhultgren) [inline src/server/api/controllers/orders.controller.js]: Nice 👍🏼

**Review by miltonhultgren — APPROVED** [review]
Nice, but I'm not sure we should use `+` instead of `parseInt`.

File: src/server/api/controllers/orders.controller.js (hunk 15 above)
Comment (JiuTak) [reply src/server/api/controllers/orders.controller.js:49]: Ow that because it cannot catch when the input is 1r(fx, number+character) or something with other alphabets 
so I came up with + so it hit the right error, not showing or matching the data

Comment (miltonhultgren) [reply src/server/api/controllers/orders.controller.js:49]: I see, you are right. 
In that case, I would recommend using regular expressions instead in the future https://stackoverflow.com/a/19715367
